### PR TITLE
Junit testing

### PR DESCRIPTION
**Pull request by SsopoWolox**
Summary

Completed test controller class creation
Completed test repository class creation
Completed build module training
Completed run module training
Completed versioning project for this version

Trello Card
https://trello.com/c/u2KO3H3a/28-junit-testing

### CODE REVIEW

File: src/main/java/wolox/training/models/Book.java [hunk 1]
```diff
@@ -19,6 +18,42 @@
...
+        this.author = author;
+        this.image = image;
+        this.title = title;
+        this.subtitle = subtitle;
+        this.publisher = publisher;
+        this.year = year;
+        this.pages = pages;
+        this.isbn = isbn;
+        this.users = users;
+    }
```
Comment (jeymarsonwolox) [inline src/main/java/wolox/training/models/Book.java]: don't need two constructors, remove one and send `users = null` when you don't need this

File: src/test/java/wolox/training/commons/Utils.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package wolox.training.commons;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class Utils {
+
+    public static String asJsonString(final Object obj) {
```
Comment (jeymarsonwolox) [inline src/test/java/wolox/training/commons/Utils.java:7]: great!!

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    @Test
+    public void givenBook_whenUpdate_thenReturnUpdateBook() throws Exception {
+        when(mockBookRepository.findById(1L)).thenReturn(java.util.Optional.ofNullable(mockBook));
+        when(mockBookRepository.save(any())).thenReturn(mockBook);
+        mvc.perform(MockMvcRequestBuilders.put("/api/books/{id}", 1)
+                .accept(MediaType.APPLICATION_JSON_VALUE)
+                .content(
+                        "{\"id\":1,\"genre\":\"Terror\",\"author\":\"Stephen King\",\"image\":\"image2.pgn\",\"title\":\"It\",\"subtitle\":\"-\",\"publisher\":\"Viking Press\",\"year\":\"1986\",\"pages\":200,\"isbn\":\"45788865\"}")
```
Comment (jeymarsonwolox) [inline src/test/java/wolox/training/controllers/BookControllerTest.java]: extract json to a constant

File: src/test/java/wolox/training/repositories/UserRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        mockBook = new Book();
+        mockBook.setAuthor("Stephen King");
+        mockBook.setGenre("Terror");
+        mockBook.setImage("image2.pgn");
+        mockBook.setIsbn("45788865");
+        mockBook.setPages(200);
+        mockBook.setPublisher("Viking Press");
+        mockBook.setSubtitle("-");
+        mockBook.setTitle("It");
+        mockBook.setYear("1986");
```
Comment (jeymarsonwolox) [inline src/test/java/wolox/training/repositories/UserRepositoryTest.java]: you are repeating this many times, extract to a new class eg `BookFactory`

**Review by jeymarsonwolox — CHANGES_REQUESTED** [review]

**Review by jeymarsonwolox — DISMISSED** [review]

File: src/test/java/wolox/training/repositories/BookRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    public void givenBook_whenSave_returnPersistenceException() {
+        Exception exception = Assertions.assertThrows(PersistenceException.class, () ->
+                entityManager.merge(new Book(1L, "Terror", null, "image2.pgn", "It", "-",
+                        "Viking Press", "1986", 200, "45788865", null)));
+        Assertions.assertTrue(exception.getCause() instanceof PropertyValueException);
+    }
+
+    @Test
+    public void givenBook_whenIsNotNull_returnOk() {
+        org.assertj.core.api.Assertions.assertThat(BookFactory.getBook().getAuthor()).isNotNull();
```
Comment (martinremis) [inline src/test/java/wolox/training/repositories/BookRepositoryTest.java]: Add this as an import in order to avoid the complete method path.

**Review by martinremis — CHANGES_REQUESTED** [review]

File: src/test/java/wolox/training/repositories/BookRepositoryTest.java (hunk 5 above)
Comment (martinremis) [reply src/test/java/wolox/training/repositories/BookRepositoryTest.java]: Do this for all the uses of `org.assertj.core.api.Assertions.assertThat(`

**Review by martinremis — APPROVED** [review]

**Review by jeymarsonwolox — APPROVED** [review]

**Pull request merged**